src/wrappers/themis/python/pythemis/__init__.py

```diff
@@ -14,10 +14,91 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+
+import os
+import sys
+import warnings
+from ctypes import cdll
+from ctypes.util import find_library
+
+
+def _is_linux():
+    return sys.platform.startswith('linux')
+
+
+def _is_macos():
+    return sys.platform.startswith('darwin')
+
+
+def _load_themis():
+    try:
+        if _is_linux():
+            return cdll.LoadLibrary('libthemis.so.0')
+        if _is_macos():
+            return cdll.LoadLibrary('libthemis.0.dylib')
+    except Exception as exception:
+        warnings.warn("""failed to load the canonical Themis Core library
+
+Exception: {}
+
+Proceeding to find 'themis' library in standard paths.
+This might cause ABI mismatch and crash the process.
+""".format(exception),
+            category=RuntimeWarning)
+
+    themis_path = find_library('themis')
+    # find_library() returns None on failure and LoadLibrary() would
+    # happily 'load' that. We kinda need the library, make it an error.
+    if not themis_path:
+        raise RuntimeError('failed to locate Themis Core library')
+    return cdll.LoadLibrary(themis_path)
+
+
+def _default_themis_paths():
+    if _is_linux():
+        return [
+            '/usr/local/lib',       # default "make install" path
+        ]
+    if _is_macos():
+        return [
+            '/usr/local/lib',       # default "make install" path, Homebrew path on x86
+            '/opt/homebrew/lib',    # Homebrew path on M1
+        ]
+    return []
+
+
+class _extra_paths():
+    def __init__(self, paths):
+        self.paths = paths
+
+    def __enter__(self):
+        if _is_linux() or _is_macos():
+            if _is_macos():
+                self.varname = 'DYLD_LIBRARY_PATH'
+            else:
+                self.varname = 'LD_LIBRARY_PATH'
+            self.old_paths = os.environ.get(self.varname, None)
+            if self.old_paths:
+                new_paths = self.old_paths + ':' + ':'.join(self.paths)
+            else:
+                new_paths = ':'.join(self.paths)
+            os.environ[self.varname] = new_paths
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if _is_linux() or _is_macos():
+            if self.old_paths:
+                os.environ[self.varname] = self.old_paths
+            else:
+                del os.environ[self.varname]
+
+
 __all__ = [
     "scell",
     "scomparator",
     "skeygen",
     "smessage",
     "ssession",
 ]
+
+with _extra_paths(_default_themis_paths()):
```

Comment:
why don't use this context manager and all these wrappings in the `_load_themis`? do we have any case when we want to provide a way to load themis library without such wrapper?

Reply:
I don't think at the moment we'd want to not have that context.

I don't think there is a use case for `_load_themis` outside of this module either.

By wrapping the context outside of the call I was mostly minimizing the diff, and keeping things separated instead of piling everything into one function.

Reply:
yes, but my question is why we don't want to encapsulate it in one function and separate it into several. I understand the reason if we want to use them separately in the other places

Reply:
Readability. It's easier to read
```python
with _extra_paths(_default_themis_paths()):
    themis = _load_themis()
```
than to have the entire body of `_load_themis` intended an extra level because of that `with`. It already looks weird because of those heredoc strings that require no indentation. I don't want to make it even more weird.